src/transformers/utils/quantizers.py

```diff
@@ -0,0 +1,715 @@
+                "requirements of `bitsandbytes` to enable model loading in 8-bit or 4-bit. "
+                "Pass your own torch_dtype to specify the dtype of the remaining non-linear layers or pass"
+                " torch_dtype=torch.float16 to remove this warning.",
+                torch_dtype,
+            )
+            torch_dtype = torch.float16
+        return torch_dtype
+
+    def update_device_map(self, device_map):
+        """called right after quantizer init to fig initial device map"""
+        if device_map is None:
+            device_map = {"": torch.cuda.current_device()}
+            logger.info(
+                "The device_map was not initialized."
+                "Setting device_map to `%s`."
+                "If you want to use the model for inference, please set device_map ='auto'",
+                str(torch.cuda.current_device()),
+            )
+        return device_map
+
+    def get_special_dtypes_update(self, model, torch_dtype):
+        return {
+            name: torch_dtype
+            for name, _ in model.named_parameters()
+            if any(m in name for m in self.modules_to_not_convert)
+        }
+
+    def adjust_max_memory(self, max_memory):
+        # need more space for buffers that are created during quantization
+        max_memory = {key: val * 0.90 for key, val in max_memory.items()}
+        return max_memory
+
+    def validate_device_map(self, device_map):
+        """called after infer_auto_device_map()"""
+        # TODO: consider combining with  .update_device_map()
+        # TODO: consider combining with  .adjust_max_memory()
+        device_map_without_lm_head = {
+            key: device_map[key] for key in device_map.keys() if key not in self.modules_to_not_convert
+        }
+        if "cpu" in device_map_without_lm_head.values() or "disk" in device_map_without_lm_head.values():
+            raise ValueError(
+                """
+                Some modules are dispatched on the CPU or the disk. Make sure you have enough GPU RAM to fit the
+                quantized model. If you want to dispatch the model on the CPU or the disk while keeping these modules
+                in 32-bit, you need to set `load_in_8bit_fp32_cpu_offload=True` and pass a custom `device_map` to
+                `from_pretrained`. Check
+                https://huggingface.co/docs/transformers/main/en/main_classes/quantization#offload-between-cpu-and-gpu
+                for more details.
+                """
+            )
+
+    def is_model_trainable(self, model=None):
+        return version.parse(importlib.metadata.version("bitsandbytes")) >= version.parse("0.37.0")
+
+    def process_model_before_weight_loading(self, model, **kwargs):
+        super().process_model_before_weight_loading(model)
+
+    def process_model_after_weight_loading(self, model):
+        super().process_model_after_weight_loading(model)
+
+
+class BnbEightBitHFQuantizer(BnbHFQuantizer):
```

Comment:
Do you think it would make sense to just declare `Bnb8bitHFQuantizer` and `Bnb4bitHFQuantizer` since only 2 methods are using `super()` - what do you think?

Reply:
@younesbelkada, Besides these 2 methods with `super()` there are 8 other shared methods in `BnbHFQuantizer`. They correspond places in the original code with condition `load_in_8bit or load_in_4bit`.
Actually there is some common code in `create_quantized_param()` and `process_model_before_weight_loading()` to be moved into `BnbHFQuantizer` class.